model_signing/signing/in_toto.py

```diff
@@ -123,3 +124,133 @@ def from_manifest(cls, manifest: manifest_module.Manifest) -> Self:
+    manifest: manifest_module.Manifest,
+    *,
+    predicate_type: str,
+    predicate_top_level_name: str,
+):
+    """Converts manifest descriptors to an in-toto statement with payload.
+
+    Args:
+        manifest: The manifest to extract the descriptors from. Assumed valid.
+        predicate_type: The predicate_type to use in the in-toto statement.
+        predicate_top_level_name: Name to use in the payload for the array of
+          the subjects.
+    """
+    hasher = memory.SHA256()
+    subjects = []
+    for descriptor in manifest.resource_descriptors():
+        hasher.update(descriptor.digest.digest_value)
+        subjects.append({
+            "name": descriptor.identifier,
+            "digest": descriptor.digest.digest_hex,
+            "algorithm": descriptor.digest.algorithm,
+        })
+
+    digest = {"sha256": hasher.compute().digest_hex}
+    descriptor = statement.ResourceDescriptor(digest=digest).pb
+
+    return statement.Statement(
+        subjects=[descriptor],
+        predicate_type=predicate_type,
+        predicate={predicate_top_level_name: subjects},
+    )
+
+
+class DigestOfDigestsIntotoPayload(IntotoPayload):
+    """In-toto payload where the subject is a digest of digests of model files.
+
+    This payload is supposed to be used for manifests where every file in the
+    model is matched with a digest. Because existing tooling only supports
+    established hashing algorithms, we record every such digest in the predicate
+    part and compute a hash for the subject by using sha256 on the concatenation
+    of the file hashes. To ensure determinism, the hashes are sorted
+    alphabetically by filename.
+
+    Example:
+    ```json
+    {
+      "_type": "https://in-toto.io/Statement/v1",
+      "subject": [
+        {
+          "digest": {
+            "sha256": "18b5a4..."
+          }
+        }
+      ],
+      "predicateType": "https://model_signing/DigestOfDigests/v0.1",
+      "predicate": {
+        "files": [
+          {
+            "digest": "6efa14...",
+            "algorithm": "file-sha256",
+            "name": "d0/d1/d2/d3/d4/f0"
+          },
+          {
+            "digest": "a9bc14...",
+            "algorithm": "file-sha256",
+            "name": "d0/d1/d2/d3/d4/f1"
+          },
+          {
+            "digest": "5f597e...",
+            "algorithm": "file-sha256",
+            "name": "d0/d1/d2/d3/d4/f2"
+          },
+          {
+            "digest": "eaf677...",
+            "algorithm": "file-sha256",
+            "name": "d0/d1/d2/d3/d4/f3"
+          }
+        ]
+      }
+    }
+    ```
+
+    A missing predicate, or a predicate for which an entry does not have valid
+    name, digest, or algorithm should be considered invalid and fail integrity
+    verification.
+
+    See also https://github.com/sigstore/sigstore-python/issues/1018.
+    """
+
+    predicate_type: Final[str] = "https://model_signing/DigestOfDigests/v0.1"
```

Comment:
I was curious if it mattered that this wasn't resolvable. [The answer is apparently no](https://github.com/in-toto/attestation/blob/main/spec/v1/field_types.md#typeuri), but should we namespace this any more to avoid collisions?


> SHOULD resolve to a human-readable description, but MAY be unresolvable. SHOULD include a version number to allow for revisions.
> 
> TypeURIs are not registered. The natural namespacing of URIs is sufficient to prevent collisions.

Reply:
Likely yes. Right now all these constants are just for testing. When we standardize, we will pick sensible normal values for them.

Thank you for flagging and the review!